pallets/rc-migrator/src/staking/message.rs

```diff
@@ -200,318 +129,629 @@ pub struct StakingValues<Balance> {
+	PartialEq,
+	Eq,
+	Clone,
+	Encode,
+	Decode,
+	DecodeWithMemTracking,
+	RuntimeDebug,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+pub struct PortableActiveEraInfo {
+	pub index: EraIndex,
+	pub start: Option<u64>,
+}
+
+impl IntoPortable for pallet_staking::ActiveEraInfo {
+	type Portable = PortableActiveEraInfo;
+
+	fn into_portable(self) -> Self::Portable {
+		PortableActiveEraInfo { index: self.index, start: self.start }
+	}
+}
+
+impl Into<pallet_staking_async::ActiveEraInfo> for PortableActiveEraInfo {
+	fn into(self) -> pallet_staking_async::ActiveEraInfo {
+		pallet_staking_async::ActiveEraInfo { index: self.index, start: self.start }
+	}
+}
+
+#[derive(
+	PartialEq,
+	Eq,
+	Clone,
+	Encode,
+	Decode,
+	DecodeWithMemTracking,
+	RuntimeDebug,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+pub enum PortableForcing {
+	NotForcing,
+	ForceNew,
+	ForceNone,
+	ForceAlways,
+}
+
+// Forcing: RC -> Portable
+impl IntoPortable for pallet_staking::Forcing {
+	type Portable = PortableForcing;
+
+	fn into_portable(self) -> Self::Portable {
+		match self {
+			pallet_staking::Forcing::NotForcing => PortableForcing::NotForcing,
+			pallet_staking::Forcing::ForceNew => PortableForcing::ForceNew,
+			pallet_staking::Forcing::ForceNone => PortableForcing::ForceNone,
+			pallet_staking::Forcing::ForceAlways => PortableForcing::ForceAlways,
+		}
+	}
+}
+
+// Forcing: Portable -> AH
+impl Into<pallet_staking_async::Forcing> for PortableForcing {
+	fn into(self) -> pallet_staking_async::Forcing {
+		match self {
+			PortableForcing::NotForcing => pallet_staking_async::Forcing::NotForcing,
+			PortableForcing::ForceNew => pallet_staking_async::Forcing::ForceNew,
+			PortableForcing::ForceNone => pallet_staking_async::Forcing::ForceNone,
+			PortableForcing::ForceAlways => pallet_staking_async::Forcing::ForceAlways,
+		}
+	}
+}
+
+#[derive(
+	PartialEq,
+	Eq,
+	Clone,
+	Encode,
+	Decode,
+	DecodeWithMemTracking,
+	RuntimeDebug,
+	TypeInfo,
+	MaxEncodedLen,
+)]
+pub struct PortableStakingLedger {
+	pub stash: AccountId32,
+	pub total: u128,
+	pub active: u128,
+	pub unlocking: BoundedVec<PortableUnlockChunk, ConstU32<100>>, /* 100 is an upper bound TODO
+	                                                                * @kianenigma review */
+}
+
+// StakingLedger: RC -> Portable
+impl<T: Config> IntoPortable for pallet_staking::StakingLedger<T> {
+	type Portable = PortableStakingLedger;
+
+	fn into_portable(self) -> Self::Portable {
+		// TODO @kianenigma what to do with `legacy_claimed_rewards`?
+		//defensive_assert!(self.legacy_claimed_rewards.is_empty());
+
+		PortableStakingLedger {
+			stash: self.stash,
+			total: self.total,
+			active: self.active,
+			unlocking: self
 				.unlocking
 				.into_iter()
-				.map(pallet_staking::UnlockChunk::intoAh)
+				.map(IntoPortable::into_portable)
 				.collect::<Vec<_>>()
 				.defensive_truncate_into(),
-			// legacy_claimed_rewards not migrated
-			controller: ledger.controller,
+			// TODO @kianenigma controller is ignored, right?
```

Comment:
Right. `Controller` is the storage key for the ledger. Its included as a field in the struct for decoration but isnt actually stored in the value itself.

Reply:
ref: https://github.com/paritytech/polkadot-sdk/blob/599ff333d2e930db8ed96fc1cba6022b74d9c06f/substrate/frame/staking-async/src/ledger.rs#L106